### PR TITLE
DAOS-11955 pool: Ensure a PS is inside pool

### PR DESCRIPTION
**Pull request by liw**
It was found that a PS leader may enter ds_pool_plan_svc_reconfs with
itself being an undesirable replica. This may lead to an assertion
failure at "move n replicas from undesired to to_remove" in
ds_pool_plan_svc_reconfs. Moreover, such a PS leader may be outside of
the pool group, making it incapable of performing many duties that
involve collective communication.

This patch therefore ensures that a PS leader will remove undesirable PS
replicas synchronously before committing a pool map modification that
introduces new undesirable PS replicas. (If we were to keep an
undesirable PS replica, it might become a PS leader.)

  - Extend and clean up pool_svc_sched.
      * Allow pool_svc_reconf_ult to return an error, so that we can
	fail a pool map modification if its synchronous PS replica
        removal fails.
      * Allow pool_svc_reconf_ult to get an argument, so that we can
	tell pool_svc_reconf_ult whether we want a synchronous
        remove-only run or an asyncrhonous add-remove run.
      * Move pool_svc_sched.{psc_svc_rf,psc_force_notify} up to
	pool_svc.
  - Prevent pool_svc_step_up_cb from canceling in-progress
    reconfigurations by comparing pool map versions for which the
    reconfigurations are scheduled.
  - Rename POOL_GROUP_MAP_STATUS to POOL_GROUP_MAP_STATES so that we are
    consistent with the pool_map module.

Signed-off-by: Li Wei <wei.g.li@intel.com>
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by liw** [comment]
`Features: pool` runs have only encountered two known failures; no new ones.

**Review by kccain — CHANGES_REQUESTED** [review]
just a minor correction to a comment pretty much. Looks good.

File: src/pool/srv_pool.c [hunk 1]
```diff
@@ -5918,6 +6048,36 @@ pool_svc_update_map_internal(struct pool_svc *svc, unsigned int opc,
 	if (map_version == map_version_before)
 		D_GOTO(out_map, rc = 0);
 
+	/*
+	 * If the map modification affects myself, leave it to a new PS leader
+	 * if there's another PS replica, or reject it.
+	 */
+	rc = pool_map_find_nodes(map, dss_self_rank(), &node);
```
Comment (liuxuezhao) [inline src/pool/srv_pool.c:6055]: not very sure if using "dss_self_rank()" is always correct, for example if the pool is created over ranks 1,3,5,7 out of total 8 ranks (0 - 7), the 4 ranks in the pool's sub-group will with logical rank 0 - 3?
But if with problem there are some other places also with the issue.

**Review by liuxuezhao — APPROVED** [review]
LGTM, but I am not familiar with those detailed logic.

File: src/pool/srv_pool.c (hunk 1 above)
Comment (liw) [reply src/pool/srv_pool.c:6055]: Thanks, Xuezhao. Do you mean I should have used `pool_map_find_node_by_rank` instead? I think you're right.

Ranks in pool map are PG ranks, not SG ranks. If I'm not mistaken, I think using `dss_self_rank` is correct.

Comment (liuxuezhao) [reply src/pool/srv_pool.c]: OK, using dss_self_rank is correct, thx. And your change to pool_map_find_node_by_rank() is also correct although it is not my original intended suggestion, but happened to find an issue, :)

**Review by liuxuezhao — APPROVED** [review]

**Pull request merged**